navigation/src/androidMain/kotlin/com/freeletics/khonshu/navigation/NavigationSetup.kt

```diff
@@ -1,22 +1,76 @@
-package com.freeletics.khonshu.navigation.internal
+package com.freeletics.khonshu.navigation
 
 import android.app.Activity
+import android.content.Context
+import android.content.ContextWrapper
+import android.os.Parcelable
+import androidx.activity.compose.LocalOnBackPressedDispatcherOwner
+import androidx.activity.compose.rememberLauncherForActivityResult
 import androidx.activity.result.ActivityResultLauncher
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.DisposableEffect
+import androidx.compose.runtime.LaunchedEffect
+import androidx.compose.runtime.ProvidableCompositionLocal
+import androidx.compose.runtime.staticCompositionLocalOf
+import androidx.compose.ui.platform.LocalContext
+import androidx.compose.ui.platform.LocalLifecycleOwner
 import androidx.core.app.ActivityCompat
 import androidx.lifecycle.Lifecycle
 import androidx.lifecycle.flowWithLifecycle
-import com.freeletics.khonshu.navigation.ActivityResultRequest
-import com.freeletics.khonshu.navigation.ContractResultOwner
-import com.freeletics.khonshu.navigation.NavEventNavigator
-import com.freeletics.khonshu.navigation.NavigationResultRequest
-import com.freeletics.khonshu.navigation.PermissionsResultRequest
-import com.freeletics.khonshu.navigation.PermissionsResultRequest.PermissionResult
+import com.freeletics.khonshu.navigation.internal.InternalNavigationApi
+import com.freeletics.khonshu.navigation.internal.NavEvent
+import com.freeletics.khonshu.navigation.internal.NavigationExecutor
 import kotlinx.coroutines.Dispatchers
 import kotlinx.coroutines.withContext
 import kotlinx.parcelize.Parcelize
+import org.jetbrains.annotations.VisibleForTesting
 
-@InternalNavigationApi
-public suspend fun NavEventNavigator.collectAndHandleNavEvents(
+/**
+ * Sets up the [NavEventNavigator] inside the current composition so that it's events
+ * are handled while the composition is active.
+ */
+@Composable
+public fun NavigationSetup(navigator: NavEventNavigator) {
+    val executor = LocalNavigationExecutor.current
+    val lifecycleOwner = LocalLifecycleOwner.current
+    val context = LocalContext.current
+
+    val activityLaunchers = navigator.activityResultRequests.associateWith {
+        rememberResultLaunchers(it, context)
+    }
+
+    navigator.navigationResultRequests.forEach {
+        LaunchedEffect(executor, it) {
+            executor.collectAndHandleNavigationResults(it)
+        }
+    }
+
+    val backDispatcher = LocalOnBackPressedDispatcherOwner.current!!.onBackPressedDispatcher
```

Comment:
What do you think about adding explicit failure here?

```suggestion
    val backDispatcher = LocalOnBackPressedDispatcherOwner.current?.onBackPressedDispatcher
        ?: throw IllegalStateException("Root Composable has to be hosted in androidx.activity.ComponentActivity")

```